src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -1274,83 +1468,44 @@ private void WriteHeaders(HttpRequestMessage request, ref ArrayBuffer headerBuff
                 {
-                    if (_pool.EnableMultipleHttp2Connections)
-                    {
-                        throw new HttpRequestException(null, null, RequestRetryType.RetryOnStreamLimitReached);
-                    }
+                    // We have exhausted StreamIds. Shut down the connection.
+                    Shutdown();
+                }
 
-                    if (HttpTelemetry.Log.IsEnabled())
-                    {
-                        // Only log Http20RequestLeftQueue if we spent time waiting on the queue
-                        ValueStopwatch stopwatch = ValueStopwatch.StartNew();
-                        await _concurrentStreams.RequestCreditAsync(1, cancellationToken).ConfigureAwait(false);
-                        HttpTelemetry.Log.Http20RequestLeftQueue(stopwatch.GetElapsedTime().TotalMilliseconds);
-                    }
-                    else
-                    {
-                        await _concurrentStreams.RequestCreditAsync(1, cancellationToken).ConfigureAwait(false);
-                    }
+                if (_shutdown)
+                {
+                    // The connection has shut down. Throw a retryable exception so that this request will be handled on another connection.
+                    ThrowRetry(SR.net_http_server_shutdown);
                 }
-            }
-            catch (ObjectDisposedException)
-            {
-                // We have race condition between shutting down and initiating new requests.
-                // When we are shutting down the connection (e.g. due to receiving GOAWAY, etc)
-                // we will wait until the stream count goes to 0, and then we will close the connetion
-                // and perform clean up, including disposing _concurrentStreams.
-                // So if we get ObjectDisposedException here, we must have shut down the connection.
-                // Throw a retryable request exception if this is not result of some other error.
-                // This will cause retry logic to kick in and perform another connection attempt.
-                // The user should never see this exception.  See similar handling below.
-                // Throw a retryable request exception if this is not result of some other error.
-                // This will cause retry logic to kick in and perform another connection attempt.
-                // The user should never see this exception.  See also below.
-                lock (SyncObject)
+
+                if (_streamsInUse > _maxConcurrentStreams)
                 {
-                    Debug.Assert(_disposed || _lastStreamId != -1);
-                    Debug.Assert(_httpStreams.Count == 0);
-                    ThrowShutdownException();
-                    throw; // unreachable
+                    // The server must have sent a downward adjustment to SETTINGS_MAX_CONCURRENT_STREAMS, so our previous stream reservation is no longer valid.
```

Comment:
We could also avoid ACKing the settings frame until we're at/under the limit including reservations.

Reply:
We could. Doing that seems nontrivial, so for now at least I'm just keeping the existing behavior.

I agree we should consider this in the future though.